File: db/migrate-pg-to-sqlite.py
#!/usr/bin/env python3
"""
Migrate PostgreSQL biluppgifter database → SQLite

Columns are auto-detected from PostgreSQL — no hardcoded schema assumptions.
The script also prints a column map so you can see exactly what landed in SQLite.

Usage:
    python3 db/migrate-pg-to-sqlite.py [options]

Options:
    --pg-host      PG host        (env: DB_HOST,        default: localhost)
    --pg-port      PG port        (env: DB_PORT,        default: 5432)
    --pg-db        PG database    (env: DB,             default: biluppgifter)
    --pg-user      PG user        (env: DB_USER)
    --pg-password  PG password    (env: DB_USER_PASSWD)
    --output       SQLite path    (env: DB_PATH,        default: ./biluppgifter.db)
    --dry-run      Print schema + row counts, don't write SQLite

Dependencies:
    pip install psycopg2-binary
    (sqlite3 is part of the Python standard library)
"""

import argparse
import os
import sqlite3
import sys
from pathlib import Path

try:
    import psycopg2
    import psycopg2.extras
except ImportError:
    sys.exit("Missing dependency: run  pip install psycopg2-binary")


# ── Tables to migrate (in FK-safe order) ─────────────────────────────────────

TABLES = [
    "users",
    "cars",
    "owners",
    "owner_cars",
    "lists",
    "list_items",
    "link_to_merinfo",
]

# PostgreSQL type → SQLite affinity
PG_TO_SQLITE = {
    "integer":           "INTEGER",
    "bigint":            "INTEGER",
    "smallint":          "INTEGER",
    "serial":            "INTEGER",
    "bigserial":         "INTEGER",
    "boolean":           "INTEGER",
    "real":              "REAL",
    "double precision":  "REAL",
    "numeric":           "REAL",
    "decimal":           "REAL",
    "text":              "TEXT",
    "varchar":           "TEXT",
    "character varying": "TEXT",
    "char":              "TEXT",
    "character":         "TEXT",
    "date":              "TEXT",
    "timestamp":         "TEXT",
    "timestamp without time zone": "TEXT",
    "timestamp with time zone":    "TEXT",
    "bytea":             "BLOB",
}


# ── Helpers ───────────────────────────────────────────────────────────────────

def ok(msg):    print(f"  \u2713 {msg}")
def info(msg):  print(f"  \u00b7 {msg}")
def warn(msg):  print(f"  ! {msg}")
def header(msg):print(f"\n{msg}")


def pg_columns(cur, table):
    """Return list of column dicts for a table, using a plain (tuple) cursor."""
    cur.execute("""
        SELECT column_name,
               data_type,
               is_nullable,
               column_default
        FROM information_schema.columns
        WHERE table_schema = 'public'
          AND table_name   = %s
        ORDER BY ordinal_position
    """, (table,))
    result = []
    for col_name, pg_type, nullable, default in cur.fetchall():
        sqlite_type = PG_TO_SQLITE.get(pg_type.lower(), "TEXT")
        result.append({
            "name":     col_name,
            "type":     sqlite_type,
            "nullable": nullable == "YES",
            "default":  default,
        })
    return result


def pg_primary_key(cur, table):
    """Return list of PK column names for a table, using a plain (tuple) cursor."""
    cur.execute("""
        SELECT kcu.column_name
        FROM information_schema.table_constraints tc
        JOIN information_schema.key_column_usage kcu
          ON tc.constraint_name = kcu.constraint_name
         AND tc.table_schema    = kcu.table_schema
        WHERE tc.constraint_type = 'PRIMARY KEY'
          AND tc.table_schema    = 'public'
          AND tc.table_name      = %s
        ORDER BY kcu.ordinal_position
    """, (table,))
    return [row[0] for row in cur.fetchall()]


def build_create_table(table, columns, pk_cols):
    """Generate CREATE TABLE IF NOT EXISTS DDL for SQLite."""
    col_defs = []
    for c in columns:
        parts = [f'"{c["name"]}"', c["type"]]
        if not c["nullable"]:
            parts.append("NOT NULL")
        col_defs.append(" ".join(parts))

    if pk_cols:
        quoted = ", ".join(f'"{c}"' for c in pk_cols)
        col_defs.append(f"PRIMARY KEY ({quoted})")

    body = ",\n    ".join(col_defs)
    return f'CREATE TABLE IF NOT EXISTS "{table}" (\n    {body}\n);'


def copy_table(pg_rows, sqlite_conn, table, col_names):
    if not pg_rows:
        info(f"{table}: 0 rows (skipped)")
        return

    quoted_cols  = ", ".join(f'"{c}"' for c in col_names)
    placeholders = ", ".join("?" * len(col_names))
    sql = f'INSERT OR IGNORE INTO "{table}" ({quoted_cols}) VALUES ({placeholders})'

    rows = [tuple(row[c] for c in col_names) for row in pg_rows]
    sqlite_conn.executemany(sql, rows)
    ok(f"{table}: {len(rows)} rows copied")


# ── Static views (reference well-known column names from the app) ─────────────
# These are created after the tables; if your column names differ the view
# creation will fail gracefully with a warning rather than aborting the migration.

VIEWS = {
    "my_view": """
        SELECT car_id, car_plate, car_model, type_name, car_color, car_year
        FROM cars
    """,
    "list_count_view": """
        SELECT l.list_id, l.list_name, l.user_id, COUNT(li.car_id) AS count
        FROM lists l
        LEFT JOIN list_items li ON l.list_id = li.list_id
        GROUP BY l.list_id, l.list_name, l.user_id
    """,
    "list_view2": """
        SELECT
            c.car_plate, c.car_model, c.type_name, c.car_color, c.car_year,
            o.owner_name, o.owner_age, o.owner_street, o.owner_postnumber,
            o.owner_city, o.owner_phone,
            lm.link,
            li.list_id, c.car_id, o.owner_id, l.user_id
        FROM list_items li
        JOIN  cars  c  ON li.car_id  = c.car_id
        JOIN  lists l  ON li.list_id = l.list_id
        LEFT JOIN owner_cars      oc ON c.car_id   = oc.car_id
        LEFT JOIN owners           o ON oc.owner_id = o.owner_id
        LEFT JOIN link_to_merinfo lm ON lm.car_id   = c.car_id
                                    AND lm.owner_id  = o.owner_id
    """,
    "in_list_no_owner": """
        SELECT DISTINCT c.car_plate, c.car_id
        FROM list_items li
        JOIN  cars c  ON li.car_id = c.car_id
        LEFT JOIN owner_cars oc ON c.car_id = oc.car_id
        WHERE oc.owner_id IS NULL
    """,
}


# ── Arg parsing ───────────────────────────────────────────────────────────────

def parse_args():
    p = argparse.ArgumentParser(description="Migrate PostgreSQL → SQLite")
    p.add_argument("--pg-host",     default=os.environ.get("DB_HOST", "localhost"))
    p.add_argument("--pg-port",     default=int(os.environ.get("DB_PORT", "5432")), type=int)
    p.add_argument("--pg-db",       default=os.environ.get("DB", "biluppgifter"))
    p.add_argument("--pg-user",     default=os.environ.get("DB_USER"))
    p.add_argument("--pg-password", default=os.environ.get("DB_USER_PASSWD"))
    p.add_argument("--output",      default=os.environ.get("DB_PATH", "biluppgifter.db"))
    p.add_argument("--dry-run",     action="store_true")
    return p.parse_args()


# ── Main ──────────────────────────────────────────────────────────────────────

def main():
    args = parse_args()

    if not args.pg_user:
        sys.exit("Error: PostgreSQL user is required (--pg-user or DB_USER env var)")

    print()
    print("PostgreSQL → SQLite migration")
    print("─────────────────────────────")
    print(f"  Source : {args.pg_user}@{args.pg_host}:{args.pg_port}/{args.pg_db}")
    print(f"  Target : {args.output}")
    print(f"  Dry run: {args.dry_run}")

    # ── Connect ──────────────────────────────────────────────────────────────
    try:
        pg = psycopg2.connect(
            host=args.pg_host, port=args.pg_port,
            dbname=args.pg_db, user=args.pg_user, password=args.pg_password,
        )
        pg.set_session(readonly=True, autocommit=True)
    except psycopg2.OperationalError as e:
        sys.exit(f"\nCould not connect to PostgreSQL: {e}")

    ok("Connected to PostgreSQL")

    meta_cur = pg.cursor()  # plain tuple cursor — required for information_schema unpacking

    # ── Introspect schema ────────────────────────────────────────────────────
    header("Detected schema:")
    table_meta = {}   # table → {columns: [...], pk: [...]}
    missing = []

    for table in TABLES:
        cols = pg_columns(meta_cur, table)
        pk   = pg_primary_key(meta_cur, table)
        if not cols:
            warn(f"{table}: table not found in PostgreSQL — will be skipped")
            missing.append(table)
            continue
        table_meta[table] = {"columns": cols, "pk": pk}
        col_summary = ", ".join(
            f'{c["name"]} ({c["type"]})' for c in cols
        )
        print(f"  {table:<20}: {col_summary}")

    # ── Fetch data ───────────────────────────────────────────────────────────
    header("Fetching data from PostgreSQL…")
    data = {}
    data_cur = pg.cursor(cursor_factory=psycopg2.extras.RealDictCursor)

    for table, meta in table_meta.items():
        col_names = [c["name"] for c in meta["columns"]]
        quoted    = ", ".join(f'"{c}"' for c in col_names)
        data_cur.execute(f'SELECT {quoted} FROM "{table}"')
        data[table] = data_cur.fetchall()
        ok(f"{table}: {len(data[table])} rows")

    data_cur.close()
    meta_cur.close()
    pg.close()
    ok("Disconnected from PostgreSQL")

    if args.dry_run:
        print("\nDry run — no SQLite file written.")
        return

    # ── Write SQLite ─────────────────────────────────────────────────────────
    header("Writing to SQLite…")

    output_path = Path(args.output)
    output_path.parent.mkdir(parents=True, exist_ok=True)

    sqlite = sqlite3.connect(output_path)
    sqlite.execute("PRAGMA journal_mode = WAL")
    sqlite.execute("PRAGMA foreign_keys = OFF")

    # Create tables
    for table, meta in table_meta.items():
        ddl = build_create_table(table, meta["columns"], meta["pk"])
        sqlite.execute(ddl)
    sqlite.commit()
    ok("Tables created")

    # Copy rows
    for table, meta in table_meta.items():
        col_names = [c["name"] for c in meta["columns"]]
        copy_table(data[table], sqlite, table, col_names)
    sqlite.commit()

    # Create views
    header("Creating views…")
    for view_name, view_sql in VIEWS.items():
        try:
            sqlite.execute(
                f'CREATE VIEW IF NOT EXISTS "{view_name}" AS {view_sql}'
            )
            ok(view_name)
        except sqlite3.OperationalError as e:
            warn(f"{view_name}: skipped — {e}")
    sqlite.commit()

    sqlite.execute("PRAGMA foreign_keys = ON")

    # ── Verify ───────────────────────────────────────────────────────────────
    header("Verification (SQLite row counts):")
    for table in table_meta:
        (count,) = sqlite.execute(f'SELECT COUNT(*) FROM "{table}"').fetchone()
        print(f"  {table:<20}: {count}")

    sqlite.close()

    print(f"\nMigration complete → {output_path.resolve()}\n")

    if missing:
        print("Tables not found in PostgreSQL (skipped):", ", ".join(missing))
        print()


if __name__ == "__main__":
    main()

File: db/migrate-pg-to-sqlite.js
#!/usr/bin/env node
/**
 * Migrate PostgreSQL biluppgifter database → SQLite
 *
 * Usage:
 *   node db/migrate-pg-to-sqlite.js [options]
 *
 * Options (all have env-var fallbacks):
 *   --pg-host      PG host        (env: DB_HOST,   default: localhost)
 *   --pg-port      PG port        (env: DB_PORT,   default: 5432)
 *   --pg-db        PG database    (env: DB,        default: biluppgifter)
 *   --pg-user      PG user        (env: DB_USER)
 *   --pg-password  PG password    (env: DB_USER_PASSWD)
 *   --output       SQLite path    (env: DB_PATH,   default: ./biluppgifter.db)
 *   --dry-run      Print row counts, don't write SQLite
 *
 * Example:
 *   node db/migrate-pg-to-sqlite.js \
 *     --pg-host localhost --pg-user myuser --pg-password secret \
 *     --output /path/to/biluppgifter.db
 *
 * Dependencies (install if not present):
 *   npm install pg better-sqlite3
 */

'use strict';

const path = require('path');
const fs   = require('fs');

// ── Argument parsing ─────────────────────────────────────────────────────────

function parseArgs() {
  const args = process.argv.slice(2);
  const opts = {};
  for (let i = 0; i < args.length; i++) {
    const key = args[i].replace(/^--/, '');
    const val = args[i + 1] && !args[i + 1].startsWith('--') ? args[++i] : true;
    opts[key] = val;
  }
  return opts;
}

const args = parseArgs();

const pgConfig = {
  host:     args['pg-host']     || process.env.DB_HOST        || 'localhost',
  port:     parseInt(args['pg-port'] || process.env.DB_PORT || '5432'),
  database: args['pg-db']       || process.env.DB             || 'biluppgifter',
  user:     args['pg-user']     || process.env.DB_USER,
  password: args['pg-password'] || process.env.DB_USER_PASSWD,
};

const sqlitePath = args['output'] || process.env.DB_PATH || path.join(process.cwd(), 'biluppgifter.db');
const dryRun     = args['dry-run'] === true || args['dry-run'] === 'true';

if (!pgConfig.user) {
  console.error('Error: PostgreSQL user is required (--pg-user or DB_USER env var)');
  process.exit(1);
}

// ── Load drivers ─────────────────────────────────────────────────────────────

let pg, Database;

try {
  pg = require('pg');
} catch {
  console.error('Missing dependency: run  npm install pg');
  process.exit(1);
}

try {
  Database = require('better-sqlite3');
} catch {
  console.error('Missing dependency: run  npm install better-sqlite3');
  process.exit(1);
}

// ── Helpers ───────────────────────────────────────────────────────────────────

function log(msg) { process.stdout.write(msg + '\n'); }
function ok(msg)  { log('  ✓ ' + msg); }
function info(msg){ log('  · ' + msg); }

async function fetchAll(client, sql) {
  const res = await client.query(sql);
  return res.rows;
}

// Copy one table: fetch from PG, bulk-insert into SQLite inside a transaction.
// The INSERT uses INSERT OR IGNORE so re-running the script is safe.
function copyTable(pgRows, sqliteDb, tableName, columns) {
  if (pgRows.length === 0) {
    info(`${tableName}: 0 rows (skipped)`);
    return;
  }

  const placeholders = columns.map(() => '?').join(', ');
  const sql = `INSERT OR IGNORE INTO ${tableName} (${columns.join(', ')}) VALUES (${placeholders})`;
  const stmt = sqliteDb.prepare(sql);

  const insertMany = sqliteDb.transaction((rows) => {
    for (const row of rows) {
      // Coerce values: PostgreSQL may return numbers/dates; keep as primitive.
      const vals = columns.map(c => {
        const v = row[c];
        if (v === null || v === undefined) return null;
        if (typeof v === 'object' && !(v instanceof Date)) return String(v);
        if (v instanceof Date) return v.toISOString();
        return v;
      });
      stmt.run(vals);
    }
  });

  insertMany(pgRows);
  ok(`${tableName}: ${pgRows.length} rows copied`);
}

// ── Main ──────────────────────────────────────────────────────────────────────

async function main() {
  log('');
  log('PostgreSQL → SQLite migration');
  log('─────────────────────────────');
  log(`  Source : ${pgConfig.user}@${pgConfig.host}:${pgConfig.port}/${pgConfig.database}`);
  log(`  Target : ${sqlitePath}`);
  log(`  Dry run: ${dryRun}`);
  log('');

  // ── Connect to PostgreSQL ────────────────────────────────────────────────
  const pgClient = new pg.Client(pgConfig);
  try {
    await pgClient.connect();
    ok('Connected to PostgreSQL');
  } catch (err) {
    console.error('Could not connect to PostgreSQL:', err.message);
    process.exit(1);
  }

  // ── Fetch all tables ─────────────────────────────────────────────────────
  log('');
  log('Fetching data from PostgreSQL…');

  const [users, cars, owners, ownerCars, lists, listItems, linkToMerinfo] = await Promise.all([
    fetchAll(pgClient, 'SELECT user_id, user_name, digest FROM users'),
    fetchAll(pgClient, 'SELECT car_id, car_plate, car_model, type_name, car_color, car_year FROM cars'),
    fetchAll(pgClient, 'SELECT owner_id, owner_name, owner_age, owner_street, owner_postnumber, owner_city, owner_phone FROM owners'),
    fetchAll(pgClient, 'SELECT owner_id, car_id FROM owner_cars'),
    fetchAll(pgClient, 'SELECT list_id, list_name, user_id FROM lists'),
    fetchAll(pgClient, 'SELECT list_id, car_id FROM list_items'),
    fetchAll(pgClient, 'SELECT owner_id, car_id, link FROM link_to_merinfo'),
  ]);

  await pgClient.end();
  ok('Disconnected from PostgreSQL');

  log('');
  log('Row counts:');
  log(`  users           : ${users.length}`);
  log(`  cars            : ${cars.length}`);
  log(`  owners          : ${owners.length}`);
  log(`  owner_cars      : ${ownerCars.length}`);
  log(`  lists           : ${lists.length}`);
  log(`  list_items      : ${listItems.length}`);
  log(`  link_to_merinfo : ${linkToMerinfo.length}`);

  if (dryRun) {
    log('');
    log('Dry run — no SQLite file written.');
    return;
  }

  // ── Prepare SQLite ───────────────────────────────────────────────────────
  log('');
  log('Writing to SQLite…');

  const sqliteDir = path.dirname(sqlitePath);
  if (!fs.existsSync(sqliteDir)) fs.mkdirSync(sqliteDir, { recursive: true });

  const sqliteDb = new Database(sqlitePath);
  sqliteDb.pragma('journal_mode = WAL');
  // Disable FK checks during bulk insert so order doesn't matter within a
  // transaction; we re-enable afterwards.
  sqliteDb.pragma('foreign_keys = OFF');

  // Apply schema (creates tables + views if they don't exist yet)
  const schemaPath = path.join(__dirname, 'schema.sql');
  if (!fs.existsSync(schemaPath)) {
    console.error('schema.sql not found at', schemaPath);
    process.exit(1);
  }
  sqliteDb.exec(fs.readFileSync(schemaPath, 'utf8'));
  ok('Schema applied');

  // ── Copy tables in FK-safe order ─────────────────────────────────────────
  // (foreign_keys is OFF so the order only matters for clarity, but we keep
  //  parent tables first as a good practice)
  copyTable(users,        sqliteDb, 'users',           ['user_id', 'user_name', 'digest']);
  copyTable(cars,         sqliteDb, 'cars',            ['car_id', 'car_plate', 'car_model', 'type_name', 'car_color', 'car_year']);
  copyTable(owners,       sqliteDb, 'owners',          ['owner_id', 'owner_name', 'owner_age', 'owner_street', 'owner_postnumber', 'owner_city', 'owner_phone']);
  copyTable(ownerCars,    sqliteDb, 'owner_cars',      ['owner_id', 'car_id']);
  copyTable(lists,        sqliteDb, 'lists',           ['list_id', 'list_name', 'user_id']);
  copyTable(listItems,    sqliteDb, 'list_items',      ['list_id', 'car_id']);
  copyTable(linkToMerinfo,sqliteDb, 'link_to_merinfo', ['owner_id', 'car_id', 'link']);

  sqliteDb.pragma('foreign_keys = ON');

  // ── Verify ───────────────────────────────────────────────────────────────
  log('');
  log('Verification (SQLite row counts):');
  for (const tbl of ['users','cars','owners','owner_cars','lists','list_items','link_to_merinfo']) {
    const { count } = sqliteDb.prepare(`SELECT COUNT(*) AS count FROM ${tbl}`).get();
    log(`  ${tbl.padEnd(20)}: ${count}`);
  }

  sqliteDb.close();

  log('');
  log(`Migration complete → ${sqlitePath}`);
  log('');
}

main().catch(err => {
  console.error('\nFatal:', err.message);
  process.exit(1);
});
